src/core/gui/XournalppCursor.cpp

```diff
@@ -515,6 +520,68 @@ auto XournalppCursor::createHighlighterOrPenCursor(int size, double alpha) -> Gd
 }
 
 
+auto XournalppCursor::createHorizontalLineCursor() -> GdkCursor* {
+    GtkWidget* theWidget = control->getWindow()->getXournal()->getWidget();
+    GdkWindow* theWindow = gtk_widget_get_window(theWidget);
+
+    gint x, y;
+    GdkDevice *mouse_device;
+
+    #if GTK_CHECK_VERSION (3,20,0)
+    GdkSeat *seat = gdk_display_get_default_seat (gdk_display_get_default ());
+    mouse_device = gdk_seat_get_pointer (seat);
+    #else
+    GdkDeviceManager *devman = gdk_display_get_device_manager (gdk_display_get_default ());
+    mouse_device = gdk_device_manager_get_client_pointer (devman);
+    #endif
+
+    gdk_window_get_device_position (theWindow, mouse_device, &x, &y, nullptr);
+
+    Layout* const layout = this->control->getWindow()->getLayout();
+    XojPageView* page = layout->getPageViewAt(x, y);
+    if (page) {
+        Range range = page->getVisiblePart();
+        double zoom = control->getZoomControl()->getZoom();
+        int width = static_cast<int>(std::round(range.getWidth() * zoom));
+        int height = 5; // Five is the default cursor size used elsewhere in the app.
+        gint dx = x - static_cast<int>(range.getX() * zoom) - page->getX();
+        if (this->currentCursorFlavour == (dx << (sizeof(gint) * 4) | width)) {
```

Comment:
I'm not sure if including the dx will be very useful, since most people will not be moving the mouse/stylus in a perfectly vertical line downwards when they use the tool.

Reply:
So, including dx in the caching decision is important because that number changes depending on where your mouse is horizontally. Because of the way that the cursor is set up, it needs to be redrawn so that the line is visibly over the page each time the cursor moves horizontally.

Honestly, I dunno how effective caching is going to be at all, so I threw together a quick way to tally how many times the cursor was cached vs redrawn in a variety of ways.

*With dx*
Randomly jiggling my Magic Trackpad up and down for a few seconds
```
Caches: 203 Redraws: 543
28% Caching Rate
```

Drawing a few random lines and moving them with my Magic Trackpad
```
Caches: 273 Redraws: 322
45% Caching Rate
```

Writing a few sentences and shapes and moving them with my Wacom Tablet
```
Caches: 686 Redraws: 704
49% Caching Rate
```

I tried to be as natural as possible, but this is _far_ from scientific. Coincidentally, my first day of class is tomorrow, so I could probably run this test during an actual lecture, though it's just syllabus week :P

Reply:
50% hit rate is a lot higher than I expected, let's leave `dx` in then.